File: clean_corpus/clean_corpus.pl
#!/usr/bin/perl

use strict;
use warnings;

use utf8;
use File::Basename;
use List::Util qw(min);
use HTML::Entities;
use open ":std", ":encoding(UTF-8)";

my $timer = time();
my $path = dirname(__FILE__) . "/";

my %CONV;
open FILE, "<:encoding(UTF-8)", $path . "char_table.tsv"  or die $!;
while(<FILE>) {
    chomp;
    my ($a, $b) = split "\t", $_;
    $CONV{$a} = $b;
}
my $CONV_MAXLEN = length((sort { length($b) <=> length($a) } keys %CONV)[0]);

open FILE, "<:encoding(UTF-8)", $ARGV[0] or die $!;
while(<FILE>) {
    my $line = $_;

    # full width characters
    for my $i (0 .. length($line) - 1) {
        my $c = ord substr $line, $i, 1;
        if ($c >= 0xFF01 and $c <= 0xFF5E) {
            substr($line, $i, 1) = chr $c - hex "0xFEE0";
        }
    }

    # mixed width CJK characters
    for (my $i = 0; $i < length($line); ) {
        my $k = 1;
        for (my $j = min($CONV_MAXLEN, length($line) - $i); $j > 0; $j--) {
            my $w = substr $line, $i, $j;
            if (exists($CONV{$w})) {
                substr($line, $i, $j) = $CONV{$w};
                $k = 0;
                last;
            }
        }
        $i++ if $k;
    }

    # HTML entities
    $line = decode_entities($line);

    # control characters
    $line =~ s/[\x{0000}-\x{001F}\x{007F}\x{0080}-\x{009F}]+/ /g;

    # whitespace characters
    $line =~ s/[\x{0020}\x{00A0}\x{2000}-\x{200B}\x{202F}\x{205F}\x{3000}]+/ /g;

    # private use area
    $line =~ s/[\x{E000}-\x{F8FF}]+/ /g;

    # byte order marks
    $line =~ s/[\x{FEFF}\x{FFFE}]+/ /g;

    # punctuation marks
    $line =~ s/(?<=\S)’(?=(d|ll|m|re|s|t|ve)\b)/'/gi;

    $line =~ s/ {2,}/ /g;
    $line =~ s/^ | $//g;

    print $line, "\n";
}

printf STDERR "%f seconds\n", time() - $timer;

File: clean_corpus/clean_corpus.py
import os
import sys
import re
import html

class corpus_cleaner():

    def __init__(self):
        self.path = (os.path.dirname(__file__) or ".") + "/"
        self.table, self.maxlen = self.load_table("char_table.tsv")

    def load_table(self, filename):
        table = dict()
        with open(self.path + filename) as fo:
            for line in fo:
                a, b = line.strip().split("\t")
                table[a] = b
        maxlen = len(max(table.keys(), key = len))
        return table, maxlen

    def clean(self, line):

        # full width characters
        line = "".join(chr(ord(c) - 0xFEE0) if "\uFF01" <= c <= "\uFF5E" else c for c in line)

        # mixed width CJK characters
        i = 0
        while i < len(line):
            for j in range(min(len(line), i + self.maxlen), i, -1):
                w = line[i:j]
                if w in self.table:
                    line = line[:i] + self.table[w] + line[j:]
                    i += len(self.table[w])
                    break
            else:
                i += 1

        # HTML entities
        line = html.unescape(line)

        # control characters
        line = re.sub("[\x00-\x1F\x7F\x80-\x9F]+", " ", line)

        # whitespace characters
        line = re.sub("[\x20\xA0\u2000-\u200F\u202F\u205F\u3000]+", " ", line)

        # private use area
        line = re.sub("[\uE000-\uF8FF]", " ", line)

        # byte order marks
        line = re.sub("[\uFEFF\uFFFE]", " ", line)

        # punctuation marks
        line = re.sub(r"(?<=\S)’(?=(d|ll|m|re|s|t|ve)\b)", "'", line, flags = re.I)

        line = re.sub(" {2,}", " ", line)
        line = line.strip()

        return line

if __name__ == "__main__":
    if len(sys.argv) not in [2, 3]:
        sys.exit("Usage: %s filename [-v]" % sys.argv[0])

    cleaner = corpus_cleaner()
    verbose = (len(sys.argv) == 3 and sys.argv[2] == "-v")

    fo = open(sys.argv[1])
    for ln, raw in enumerate(fo, 1):
        line = cleaner.clean(raw)
        if not verbose:
            print(line)
        elif raw != line:
            print("<", raw, end = "")
            print(">", line, "\n")
        if ln % 100000 == 0:
            print("%d sentence pairs" % ln, file = sys.stderr)

    fo.close()
